File: kernel/arch/x64/paging.rs
use super::asm;
use super::KERNEL_BASE_ADDR;
use crate::allocator::PAGE_POOL;
use crate::arch::utils::PhyPtr;
use crate::memory::{page_fault_handler, PageFaultFlagsBuilder, PageFlags};
use crate::utils::{PAddr, VAddr};

const PAGE_SIZE: usize = 4096;
const PAGE_PRESENT: usize = 1 << 0;
const PAGE_WRITABLE: usize = 1 << 1;
const PAGE_USER: usize = 1 << 2;
const PAGE_ENTRY_NUM: usize = 512;
const PAGE_ATTRS: u64 = 0b0111;
const KERNEL_PML4_PADDR: PhyPtr<u64> = PhyPtr::new(0x0070_0000);

pub struct PageTable {
    // The physical address points to a PML4.
    pml4: PhyPtr<u64>,
}

impl PageTable {
    pub fn new() -> PageTable {
        let pml4 = PAGE_POOL.alloc_or_panic();

        // Fill the kernel space.
        unsafe {
            pml4.copy_from(PAGE_SIZE, KERNEL_PML4_PADDR.as_vaddr(), PAGE_SIZE);
        }

        PageTable {
            pml4: PhyPtr::from_vaddr(pml4),
        }
    }

    pub fn link(&self, vaddr: VAddr, paddr: PAddr, num: usize, flags: PageFlags) {
        assert!(vaddr.as_usize() < KERNEL_BASE_ADDR);

        let mut attrs = PAGE_PRESENT as u64;
        if flags.writable() {
            attrs |= PAGE_WRITABLE as u64;
        }
        if flags.user() {
            attrs |= PAGE_USER as u64;
        }

        loop {
            unsafe {
                let mut table: *mut u64 = self.pml4.as_mut_ptr();
                let mut level = 4;
                while level > 1 {
                    let index = ((vaddr.as_usize() >> (((level - 1) * 9) + 12)) & 0x1ff) as isize;
                    if table.offset(index).read() == 0 {
                        /* The PDPT, PD or PT is not allocated. Allocate it. */
                        let alloced_paddr = PAGE_POOL.alloc_or_panic().paddr().as_usize();
                        table.offset(index).write(alloced_paddr as u64);
                    }

                    // Update attributes.
                    let new_entry = (table.offset(index).read() & !PAGE_ATTRS) | attrs;
                    table.offset(index).write(new_entry);

                    // Go into the next level paging table.
                    let next_level_addr =
                        table.offset(index).read() as usize & 0x7fff_ffff_ffff_f000;
                    table = PhyPtr::<u64>::new(next_level_addr).as_mut_ptr();
                    level -= 1;
                }

                // `table` now points to the PT.
                let mut index = ((vaddr.as_usize() >> 12) & 0x1ff) as isize;
                let mut remaining = num;
                let mut offset: u64 = 0;
                let mut paddr = paddr.as_usize() as u64;
                while remaining > 0 && index < PAGE_ENTRY_NUM as isize {
                    printk!("link: %p -> %p (flags=0x%x)", vaddr.as_usize() + offset as usize, paddr, attrs);
                    table.offset(index).write(paddr | attrs);
                    asm::invlpg(vaddr.as_usize() as u64 + offset);
                    paddr += PAGE_SIZE as u64;
                    offset += PAGE_SIZE as u64;
                    remaining -= 1;
                    index += 1;
                }

                if remaining > 0 {
                    // The page spans across multiple PTs.
                    continue;
                }

                break;
            }
        }
    }

    pub fn resolve(&self, vaddr: VAddr) -> Option<PAddr> {
        unsafe {
            let mut table: *mut u64 = self.pml4.as_mut_ptr();
            let mut level = 4;
            while level > 1 {
                let index = ((vaddr.as_usize() >> (((level - 1) * 9) + 12)) & 0x1ff) as isize;
                if table.offset(index).read() == 0 {
                    return None;
                }

                // Go into the next level paging table.
                let next_level_addr =
                    table.offset(index).read() as usize & 0x7fff_ffff_ffff_f000;
                table = PhyPtr::<u64>::new(next_level_addr).as_mut_ptr();
                level -= 1;
            }

            // `table` now points to the PT.
            let index = ((vaddr.as_usize() >> 12) & 0x1ff) as isize;
            let entry = table.offset(index).read();
            if entry == 0 {
                None
            } else {
                Some(PAddr::new(entry as usize & 0x000f_ffff_ffff_f000))
            }
        }
    }

    #[inline]
    pub(super) fn pml4(&self) -> PhyPtr<u64> {
        self.pml4
    }
}

pub(super) fn handle_page_fault(error: u64) {
    let addr = unsafe { asm::get_cr2() };
    let present = (error >> 0) & 1 == 1;
    let write = (error >> 1) & 1 == 1;
    let user = (error >> 2) & 1 == 1;
    let rsvd = (error >> 3) & 1 == 1;

    if rsvd {
        panic!("#PF: RSVD bit violation");
    }

    let requested = PageFaultFlagsBuilder::new()
        .present(present)
        .user(user)
        .write(write)
        .build();

    page_fault_handler(VAddr::new(addr as usize), requested);
}

File: kernel/arch/x64/paging.c
#include <arch.h>
#include <debug.h>
#include <memory.h>
#include <printk.h>
#include <x64/x64.h>

#define NTH_LEVEL_INDEX(level, vaddr) \
    (((vaddr) >> ((((level) -1) * 9) + 12)) & 0x1ff)
#define ENTRY_PADDR(entry) ((entry) &0x7ffffffffffff000)

void page_table_init(struct page_table *pt) {
    uint64_t *pml4 = KMALLOC(&page_arena, PAGE_SIZE);
    inlined_memcpy(pml4, from_paddr(KERNEL_PML4_PADDR), PAGE_SIZE);

    // The kernel no longer access a virtual address around 0x0000_0000. Unmap
    // the area to catch bugs (especially NULL pointer dereferences in the
    // kernel).
    pml4[0] = 0;

#ifdef DEBUG_BUILD
    asan_init_area(ASAN_VALID, pml4, PAGE_SIZE);
#endif
    pt->pml4 = (paddr_t) into_paddr(pml4);
}

void page_table_destroy(UNUSED struct page_table *pt) {
    // TODO: Traverse and kfree the page table.
}

void link_page(struct page_table *pt, vaddr_t vaddr, paddr_t paddr,
               int num_pages, uintmax_t flags) {
    ASSERT(vaddr < KERNEL_BASE_ADDR && "tried to link a kernel page");

    uint64_t attrs = PAGE_PRESENT | flags;
    while (1) {
        uint64_t *table = from_paddr(pt->pml4);
        int level = 4;
        while (level > 1) {
            int index = NTH_LEVEL_INDEX(level, vaddr);
            if (!table[index]) {
                /* The PDPT, PD or PT is not allocated. Allocate it. */
                void *page = KMALLOC(&page_arena, PAGE_SIZE);
                if (!page) {
                    PANIC("failed to allocate a page for a page table");
                }

#ifdef DEBUG_BUILD
                asan_init_area(ASAN_VALID, page, PAGE_SIZE);
#endif
                table[index] = (uint64_t) into_paddr(page);
            }

            // Update attributes.
            table[index] = ENTRY_PADDR(table[index]) | attrs;

            // Go into the next level paging table.
            table = (uint64_t *) from_paddr(ENTRY_PADDR(table[index]));
            level--;
        }

        // `table` now points to the PT.
        int index = NTH_LEVEL_INDEX(1, vaddr);
        int remaining = num_pages;
        uint64_t offset = 0;
        while (remaining > 0 && index < PAGE_ENTRY_NUM) {
            // TRACE("link: %p -> %p (flags=0x%x)", vaddr, paddr, attrs);
            table[index] = paddr | attrs;
            asm_invlpg(vaddr);
            vaddr += PAGE_SIZE;
            paddr += PAGE_SIZE;
            offset += PAGE_SIZE;
            remaining--;
            index++;
        }

        if (!remaining) {
            break;
        }

        // The page spans across multiple PTs.
    }
}

void unlink_page(struct page_table *pt, vaddr_t vaddr, int num_pages) {
    ASSERT(vaddr < KERNEL_BASE_ADDR && "tried to link a kernel page");
    int remaining = num_pages;
    while (remaining > 0) {
        uint64_t *table = from_paddr(pt->pml4);
        int level = 4;
        while (level > 1) {
            int index = NTH_LEVEL_INDEX(level, vaddr);
            ASSERT(table[index]);
            table = (uint64_t *) from_paddr(ENTRY_PADDR(table[index]));
            level--;
        }

        // `table` now points to the PT.
        int index = NTH_LEVEL_INDEX(1, vaddr);
        while (remaining > 0 && index < PAGE_ENTRY_NUM) {
            // TRACE("unlink: %p", vaddr);
            table[index] = 0;
            asm_invlpg(vaddr);
            vaddr += PAGE_SIZE;
            remaining--;
            index++;
        }
    }
}

// TODO: Rewrite this function. It looks ugly :(
paddr_t resolve_paddr_from_vaddr(struct page_table *pt, vaddr_t vaddr) {
    uint64_t *table = from_paddr(pt->pml4);
    int level = 4;
    int i = 4;
    while (i > 1) {
        int index = NTH_LEVEL_INDEX(level, vaddr);
        if (!table[index]) {
            return 0;
        }

        // Go into the next level paging table.
        table = (uint64_t *) from_paddr(ENTRY_PADDR(table[index]));
        i -= (table[index] & 0x80 /* Is bigger page? */) ? 2 : 1;
        level--;
    }

    // `table` now points to the PT (4KiB page) or PD (1MiB page).
    uint64_t entry = table[NTH_LEVEL_INDEX(level, vaddr)];
    if (!entry) {
        return 0;
    }

    uint64_t offset = (level == 2)
        ? NTH_LEVEL_INDEX(level - 1, vaddr) << 12
        : 0;
    return ENTRY_PADDR(entry) + offset;
}
